nomad/peers/peers.go

```diff
@@ -133,47 +138,170 @@ func IsNomadServer(m serf.Member) (bool, *Parts) {
+
+	// localPeers is a map of the known server peers in the local region keyed
+	// by their Raft address. Peers stored here are only those in the Alive
+	// state and is used for intra-region RPC routing and discovery.
+	localPeers map[raft.ServerAddress]*Parts
+
+	// peersLock protects access to the maps above. We use a single lock so that
+	// updates are correctly applied to all maps together.
 	peersLock sync.RWMutex
 }
 
 // NewPeerCache returns a new instance of a PeerCache ready for use.
 func NewPeerCache() *PeerCache {
 	return &PeerCache{
-		peers: make(map[string][]*Parts),
+		allPeers:   make(map[string][]*Parts),
+		alivePeers: make(map[string][]*Parts),
+		localPeers: make(map[raft.ServerAddress]*Parts),
+	}
+}
+
+func (p *PeerCache) LocalPeer(addr raft.ServerAddress) *Parts {
+	p.peersLock.RLock()
+	defer p.peersLock.RUnlock()
+	return p.localPeers[addr].Copy()
+}
+
+// LocalPeers returns a list of known alive peers in the local region.
+func (p *PeerCache) LocalPeers() []*Parts {
+	p.peersLock.RLock()
+	defer p.peersLock.RUnlock()
+
+	peers := make([]*Parts, 0, len(p.localPeers))
+
+	for _, peer := range p.localPeers {
+		peers = append(peers, peer.Copy())
+	}
+
+	return peers
+}
+
+func (p *PeerCache) LocalPeersServerInfo() []*structs.NodeServerInfo {
+	p.peersLock.RLock()
+	defer p.peersLock.RUnlock()
+
+	peers := make([]*structs.NodeServerInfo, 0, len(p.localPeers))
+
+	for _, peer := range p.localPeers {
+		peers = append(peers, &structs.NodeServerInfo{
+			RPCAdvertiseAddr: peer.RPCAddr.String(),
+			Datacenter:       peer.Datacenter,
+		})
 	}
+
+	return peers
+}
+
+// RegionNum returns the number of known regions with at least one alive peer
+// and are therfore suitable for RPC routing.
+func (p *PeerCache) RegionNum() int {
+	p.peersLock.RLock()
+	defer p.peersLock.RUnlock()
+	return len(p.alivePeers)
+}
+
+// RegionNames returns the names of all known regions with at least one alive
+// peer and are therefore suitable for RPC routing.
+func (p *PeerCache) RegionNames() []string {
+	p.peersLock.RLock()
+	defer p.peersLock.RUnlock()
+
+	names := make([]string, 0, len(p.alivePeers))
+	for name := range p.alivePeers {
+		names = append(names, name)
+	}
+
+	return names
+}
+
+// RegionPeers returns the list of known alive peers in the given region. If the
+// region is not known or has no alive peers, an empty list is returned.
+func (p *PeerCache) RegionPeers(region string) []*Parts {
+	p.peersLock.RLock()
+	defer p.peersLock.RUnlock()
+
+	numPeers := len(p.alivePeers[region])
+	if numPeers == 0 {
+		return nil
+	}
+
+	peers := make([]*Parts, 0, numPeers)
+
+	for _, peer := range p.alivePeers[region] {
+		peers = append(peers, peer.Copy())
+	}
+	return peers
 }
 
 // PeerSet adds or updates the given parts in the cache. This should be called
 // when a new peer is detected or an existing peer changes is status.
-func (p *PeerCache) PeerSet(parts *Parts) {
+func (p *PeerCache) PeerSet(parts *Parts, localRegion string) {
```

Comment:
Can we call this `UpdatePeerSet` or something like that to make it clear this is a write? When seeing it at the call site (ex. `serf.go`) it was a little unexpected.